File: fs/insertSort.fs
open System

let mutable count = 0

let rec innerISort p = function
    | x :: xs when p > x ->
        count <- count + 1
        x :: innerISort p xs
    | x :: xs -> p :: x :: xs
    | _ -> [p]

let rec insertSort = function
    | x :: xs -> innerISort x <| insertSort xs
    | _ -> []

Console.ReadLine().Split(' ')
    |> Array.map int
    |> Array.toList
    |> insertSort
    |> List.iter (printf "%d ")
printfn ""
printfn "%d" count

File: ocaml/insertSort.ml
(* #load "str.cma" *)

let rec insertSort lst = 
    match lst with
    | [] -> []
    | first :: rest -> 
            let rec insert a lst2 = 
                match lst2 with
                | [] -> [a]
                | first3 :: lst3 ->
                        if first3 < a
                        then first3 :: insert a lst3
                        else a :: first3 :: lst3
            in insert first (insertSort rest)

let startN n = 
    let readinput = List.map (fun a -> Scanf.sscanf a "%d" (fun b->b)) (Str.split (Str.regexp "[ \t]+") (read_line()))
    in List.iter (Printf.printf "%d ") (insertSort readinput)

let main = Scanf.sscanf (read_line()) "%d" (fun n -> startN n)
